### ISSUE TITLE
Forward references to derived types are not supported

### ISSUE THREAD

**Issue by klausler** [flang-compiler/f18#573]
```
type, abstract :: base
  class(extended), pointer :: p
end type
type, extends(base) :: extended
end type
```

**Comment by jeanPerier** [comment]
What about allocatable components ? I know gfortran, NAG and pgfortran accepts this with pointers and not allocatables, but I did not find the exact rule here that would differentiate allocatable and pointer components on this matter.
The only restrcition to use previously defined derived-types I found are:

7.3.2.3 point 2 says:
> A derived-type-spec in a CLASS type specifier in a type declaration statement shall specify a previously defined derived type.

Which I understand as making not yet defined derived type usage illegal regardless of the attributes in type-declaration-stmt but not in data-component-def-stmt (so it is out of scope here)

C744, specific to data-component-def-stmt, says:
> If neither the POINTER nor the ALLOCATABLE attribute is specified, the declaration-type-spec in the component-def-stmt shall specify an intrinsic type or a previously defined derived type.

Therefore, I do not fully understand why ALLOCATABLE component with forward-derived type are rejected by other compilers, except that it would make checks related to potential components more annoying to perform.

I am not asking, because I want to advocate to support allocatable here too, but because I wondered when implementing `HasEventOrLockPotentialComponent` used in  allocate checks, and if we need to support this for allocatable too, then it will currently end-up in an infinite loop.
See: https://github.com/flang-compiler/f18/blob/8c63aa94972fb3f7b88f23fa1b78f8b67ffa8732/lib/semantics/tools.cc#L383

**Comment by tskeith** [comment]
Here is another example of a forward reference in a data-component-def-stmt. It is from 7.5.4.6(8) Note 5.
```
TYPE NODE
  INTEGER :: VALUE = 0
  TYPE (NODE), POINTER :: NEXT_NODE => SENTINEL
END TYPE
TYPE(NODE), SAVE, TARGET :: SENTINEL
```

**Comment by klausler** [comment]
Jean, can this bug report be closed?

**Comment by klausler** [comment]
Forward references to derived types still block some legacy correctness tests from passing; do not close yet.

(Specifically, `class(t), pointer :: p` as a component fails when `t` is defined later.)

**Comment by klausler** [comment]
Fixed (implemented) in PR #847.
